src/EFCore.PG/Storage/Internal/NpgsqlTypeMappingSource.cs

```diff
@@ -816,6 +760,57 @@ static Type FindTypeToInstantiate(Type collectionType, Type elementType)
                 ? new NpgsqlRowValueTypeMapping(clrType)
                 : null;
 
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    protected virtual RelationalTypeMapping? FindEnumMapping(in RelationalTypeMappingInfo mappingInfo)
+    {
+        var storeType = mappingInfo.StoreTypeName;
+        var clrType = mappingInfo.ClrType;
+
+        if (clrType is not null and not { IsEnum: true, IsClass: false })
+        {
+            return null;
+        }
+
+        // Try to find an enum definition (defined by the user on their context options), based on the
+        // incoming MappingInfo's StoreType or ClrType
+        EnumDefinition? enumDefinition;
+        if (storeType is null)
+        {
+            enumDefinition = _enumDefinitions.SingleOrDefault(m => m.ClrType == clrType);
+        }
+        else
+        {
+            // TODO: Not sure what to do about quoting. Is the user expected to configure properties
+            // TODO: with a quoted (schema-qualified) store type or not?
+            var dot = storeType.IndexOf('.');
+            enumDefinition = dot is -1
+                ? _enumDefinitions.SingleOrDefault(m => m.StoreTypeName == storeType)
+                : _enumDefinitions.SingleOrDefault(m => m.StoreTypeName == storeType[(dot + 1)..] && m.StoreTypeSchema == storeType[..dot]);
+        }
+
+        if (enumDefinition is null)
+        {
+            return null;
+        }
+
+        // We now have a user-defined range definition from the context options.
+
+        // We need the following store type names:
+        // 1. The quoted type name is used in migrations, where quoting is needed
+        // 2. The unquoted type name is set on NpgsqlParameter.DataTypeName
+        var (name, schema) = (enumDefinition.StoreTypeName, enumDefinition.StoreTypeSchema);
+        return new NpgsqlEnumTypeMapping(
+            _sqlGenerationHelper.DelimitIdentifier(name, schema),
+            schema is null ? name : schema + "." + name,
+            enumDefinition.ClrType,
+            enumDefinition.Labels);
```

Comment:
Yeah, this is done in the convention which "copies" the definitions from the context options to the model ([code](https://github.com/npgsql/efcore.pg/pull/3167/files#diff-df2acb3ae97846d8bf5380ebd87406f438dfb53f815dceac75c0e4dfc3ae356cR60)), because that's the part that ends up forming the migration.

Reply:
Also adding sorting in NpgsqlMigrationsSqlGenerator, when detecting new labels (this is an improvement regardless of the other changes in this PR).